File: scala/csp/ChannelDemo.scala
/**
 * ChannelDemo.scala
 *
 * Paradigm: CSP (Communicating Sequential Processes)
 *
 * CSP channel using BlockingQueue + Thread.
 * Demonstrates producer-consumer communication via channels.
 *
 * Compare with: Go channel, Java ChannelDemo, Python channel_demo
 *
 * Compile:  scalac -d build csp/ChannelDemo.scala
 * Run:      scala -classpath build ChannelDemo
 */

import java.util.concurrent.LinkedBlockingQueue

class Channel[T](capacity: Int = 0) {
  private val queue = if (capacity > 0) new LinkedBlockingQueue[T](capacity)
                      else new LinkedBlockingQueue[T]()

  def send(item: T): Unit = queue.put(item)

  def recv(): T = queue.take()

  def trySend(item: T): Boolean = queue.offer(item)

  def tryRecv(): T = queue.poll()

  def size: Int = queue.size()
}

object ChannelDemo {
  def main(args: Array[String]): Unit = {
    println("=== CSP Channel Demo (Scala) ===\n")

    // Demo 1: Unbuffered channel
    println("--- Unbuffered Channel ---")
    val ch = new Channel[Int]()

    val producer = new Thread {
      override def run(): Unit = {
        for (i <- 0 until 5) {
          val v = i * 10
          println("  Producer sending: " + v)
          ch.send(v)
        }
      }
    }
    producer.setName("Producer")

    val consumer = new Thread {
      override def run(): Unit = {
        for (_ <- 0 until 5) {
          val v = ch.recv()
          println("  Consumer received: " + v)
          Thread.sleep(50)
        }
      }
    }
    consumer.setName("Consumer")

    producer.start()
    consumer.start()
    producer.join()
    consumer.join()
    println()

    // Demo 2: Buffered channel
    println("--- Buffered Channel (capacity=3) ---")
    val buf = new Channel[Int](3)

    val fastProd = new Thread {
      override def run(): Unit = {
        for (i <- 0 until 6) {
          buf.send(i)
          println("  Produced " + i + " (buffer=" + buf.size + "/3)")
        }
      }
    }

    val slowCons = new Thread {
      override def run(): Unit = {
        for (_ <- 0 until 6) {
          val v = buf.recv()
          println("  Consumed " + v)
          Thread.sleep(100)
        }
      }
    }

    fastProd.start()
    slowCons.start()
    fastProd.join()
    slowCons.join()
    println()

    println("=== Summary ===")
    println("Scala's BlockingQueue provides CSP-like semantics.")
    println("Like Java, Scala has no native channel type or select{},")
    println("but the Thread + Queue pattern achieves the same goal.")
  }
}

File: java/csp/ChannelDemo.java
/**
 * ChannelDemo.java
 *
 * Paradigm: CSP (Communicating Sequential Processes) via BlockingQueue
 *
 * Implements a CSP-style channel using Java's LinkedBlockingQueue.
 * Demonstrates unbuffered (rendezvous) and buffered communication
 * between producer and consumer threads.
 *
 * Compare with: Go channel, Python channel_demo.py
 *
 * Compile:  javac csp/ChannelDemo.java
 * Run:      cd csp && java ChannelDemo
 */

import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public class ChannelDemo {

    // ---------------------------------------------------------------
    // Simple channel wrapper
    // ---------------------------------------------------------------
    static class Channel<T> {
        private final BlockingQueue<T> queue;

        public Channel() {
            this(0);  // unbuffered by default
        }

        public Channel(int capacity) {
            this.queue = capacity > 0
                ? new LinkedBlockingQueue<>(capacity)
                : new LinkedBlockingQueue<>(1);
        }

        /** Send — blocks until the value is received (unbuffered) or enqueued (buffered). */
        public void send(T item) throws InterruptedException {
            queue.put(item);
        }

        /** Receive — blocks until a value is available. */
        public T recv() throws InterruptedException {
            return queue.take();
        }

        /** Non-blocking send. */
        public boolean trySend(T item) {
            return queue.offer(item);
        }

        /** Non-blocking receive. */
        public T tryRecv() {
            return queue.poll();
        }

        public int size() {
            return queue.size();
        }
    }

    // ---------------------------------------------------------------
    // Demo 1: Unbuffered channel (synchronous handoff)
    // ---------------------------------------------------------------
    static void demoUnbuffered() throws InterruptedException {
        System.out.println("=== Unbuffered Channel (Rendezvous) ===");
        Channel<Integer> ch = new Channel<>();

        Thread producer = new Thread(() -> {
            try {
                for (int i = 0; i < 5; i++) {
                    int val = i * 10;
                    System.out.println("  Producer sending: " + val);
                    ch.send(val);  // blocks until consumer receives
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "Producer");

        Thread consumer = new Thread(() -> {
            try {
                for (int i = 0; i < 5; i++) {
                    int val = ch.recv();
                    System.out.println("  Consumer received: " + val);
                    Thread.sleep(50);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "Consumer");

        producer.start();
        consumer.start();
        producer.join();
        consumer.join();
        System.out.println();
    }

    // ---------------------------------------------------------------
    // Demo 2: Buffered channel
    // ---------------------------------------------------------------
    static void demoBuffered() throws InterruptedException {
        System.out.println("=== Buffered Channel (capacity=3) ===");
        Channel<Integer> ch = new Channel<>(3);

        Thread producer = new Thread(() -> {
            try {
                for (int i = 0; i < 6; i++) {
                    ch.send(i);
                    System.out.println("  Produced " + i + " (buffer=" + ch.size() + "/3)");
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "Producer");

        Thread consumer = new Thread(() -> {
            try {
                for (int i = 0; i < 6; i++) {
                    int val = ch.recv();
                    System.out.println("  Consumed " + val);
                    Thread.sleep(100);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "Consumer");

        producer.start();
        consumer.start();
        producer.join();
        consumer.join();
        System.out.println();
    }

    // ---------------------------------------------------------------
    // Main
    // ---------------------------------------------------------------
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== CSP Channel Demo (Java) ===\n");
        System.out.println("BlockingQueue + Thread 实现 CSP 风格的通道通信\n");

        demoUnbuffered();
        demoBuffered();

        System.out.println("=== Key insight ===");
        System.out.println("Java's BlockingQueue.put/take provides the same");
        System.out.println("synchronous communication semantics as Go's channel.");
        System.out.println("The main difference: Java has no select{} statement.");
    }
}
